### PR TITLE
AIP-72: Handling `failed` TI state for `AirflowFailException` & `AirflowSensorTimeout`

### PR DESCRIPTION
**Pull request by amoghrajesh**
related: https://github.com/apache/airflow/issues/44414

We already have support for handling terminal states from the task execution side as well as the task SDK client side. (almost) and failed state is part of the terminal state.

This PR extends the task runner's run function to handle cases when we have to fail a task: `AirflowFailException, AirflowSensorTimeout`. It is functionally very similar to #44786

As part of failing a task, multiple other things also needs to be done like:
- Callbacks: which will eventually be converted to teardown tasks
- Retries: Handled in https://github.com/apache/airflow/issues/44351
- unmapping TIs: https://github.com/apache/airflow/issues/44351
- Handling task history: will be handled by https://github.com/apache/airflow/issues/44952
- Handling downstream tasks and non teardown tasks: will be handled by https://github.com/apache/airflow/issues/44951

#### End to End with Postman

1. Run airflow with breeze and run any DAG
image

2. Login to metadata DB and get the "id" for your task instance from TI table
image

3. Send a request to `fail` your task
image

Or using curl:
```
curl --location --request PATCH 'http://localhost:29091/execution/task-instances/0193cec2-f46b-7348-9c27-9869d835dc7b/state' \
--header 'Content-Type: application/json' \
--data '{
    "state": "failed",
    "end_date": "2024-10-31T12:00:00Z"
}'
```

4. Refresh back the Airflow UI to see that the task is in failed state.
image

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/api_fastapi/execution_api/routes/task_instances.py
Comment (kaxil) [inline airflow/api_fastapi/execution_api/routes/task_instances.py:207]: What about other things that we need to support, the one's linked below:

https://github.com/apache/airflow/blob/8848d6e35f6ead2cc1ed9887af8360fbd61c78c3/airflow/models/taskinstance.py#L3085-L3086

Comment (kaxil) [reply airflow/api_fastapi/execution_api/routes/task_instances.py:207]: ok I see in description: https://github.com/apache/airflow/issues/44951

Comment (kaxil) [reply airflow/api_fastapi/execution_api/routes/task_instances.py:209]: end date and duration remains

File: airflow/api_fastapi/execution_api/routes/task_instances.py
Comment (kaxil) [inline airflow/api_fastapi/execution_api/routes/task_instances.py]:
```suggestion
        query = query.values(state=ti_patch_payload.state)
```

File: task_sdk/tests/execution_time/test_supervisor.py
Comment (kaxil) [inline task_sdk/tests/execution_time/test_supervisor.py]: Don't think we need this since it is already covered by `TaskState` via -- since all we need to check is it can handle `TaskState` msg:

```py
TaskState(state=TerminalTIState.SKIPPED, end_date=timezone.parse("2024-10-31T12:00:00Z")),
```

File: task_sdk/tests/execution_time/test_task_runner.py
Comment (kaxil) [inline task_sdk/tests/execution_time/test_task_runner.py]: Replace it with `CustomOperator` and raise in execute ?

Comment (kaxil) [reply task_sdk/tests/execution_time/test_task_runner.py]: You might be able to parameterize and cover all terminal states in a single one: `skipped`, `failed`, `success`

File: tests/api_fastapi/execution_api/routes/test_task_instances.py
Comment (kaxil) [inline tests/api_fastapi/execution_api/routes/test_task_instances.py:486]: Why do we need `ceil`?

**Review by kaxil — APPROVED** [review]
On high-level it lgtm -- although please address the comments I had

File: tests/api_fastapi/execution_api/routes/test_task_instances.py
Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_task_instances.py:486]: `duration_expression_update` returns the value here as a decimal which cannot be compared for equality. See an example here: https://github.com/apache/airflow/actions/runs/12350551354/job/34464066950

It returns: `3599.999986588955`.

File: airflow/api_fastapi/execution_api/routes/task_instances.py
Comment (kaxil) [inline airflow/api_fastapi/execution_api/routes/task_instances.py:205]: tbh we don't need this `if` check I think, we can set `next_method=None, next_kwargs=None` for all `TITerminalStatePayload`

File: tests/api_fastapi/execution_api/routes/test_task_instances.py
Comment (kaxil) [inline tests/api_fastapi/execution_api/routes/test_task_instances.py]:
```suggestion
            task_id="test_ti_update_state_to_failed_table_check",
```

File: tests/api_fastapi/execution_api/routes/test_task_instances.py
Comment (kaxil) [inline tests/api_fastapi/execution_api/routes/test_task_instances.py]:
```suggestion
```

File: airflow/api_fastapi/execution_api/routes/task_instances.py
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py:205]: I was overriding that in a later PR: https://github.com/apache/airflow/pull/44981

File: tests/api_fastapi/execution_api/routes/test_task_instances.py
Comment (kaxil) [reply tests/api_fastapi/execution_api/routes/test_task_instances.py:486]: @amoghrajesh Should be solved by https://github.com/apache/airflow/pull/45002

Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_task_instances.py:486]: Sqlite is only used for testing related activities right? Should we have to handle it?

Comment (kaxil) [reply tests/api_fastapi/execution_api/routes/test_task_instances.py:486]: Yeah, we do need to handle it -- also used for all the dev & getting started work

Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_task_instances.py:486]: Got it, makes sense to handle it.

File: airflow/api_fastapi/execution_api/routes/task_instances.py
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py]: ACK. Handled. Although I think the precommit made that change

File: task_sdk/tests/execution_time/test_supervisor.py
Comment (amoghrajesh) [reply task_sdk/tests/execution_time/test_supervisor.py]: Yea that's true but I kept this change to be a little explicit about which state it is going through explicitly. Would it be nice to just create a loop inside params about this instead or nuke it entirely?

File: task_sdk/tests/execution_time/test_task_runner.py
Comment (amoghrajesh) [reply task_sdk/tests/execution_time/test_task_runner.py]: That was not the intention here. In task_runner tests I am trying to instead create situations where certain `exceptions` translate to certain states. Here for example, the `AirflowFailException` and `AirflowSensorTimeout` lead to failed states.

Comment (amoghrajesh) [reply task_sdk/tests/execution_time/test_task_runner.py]: I can do it using CustomOperator though, which I will do and push

File: airflow/api_fastapi/execution_api/routes/task_instances.py
Comment (amoghrajesh) [reply airflow/api_fastapi/execution_api/routes/task_instances.py]: Alright, handled.

File: task_sdk/tests/execution_time/test_task_runner.py
Comment (amoghrajesh) [reply task_sdk/tests/execution_time/test_task_runner.py]: Handled in dcb31de9c8012872db7866d90a7d766ec313d55f

File: task_sdk/tests/execution_time/test_supervisor.py
Comment (kaxil) [reply task_sdk/tests/execution_time/test_supervisor.py]: I think nuking it is fine imo

File: tests/api_fastapi/execution_api/routes/test_task_instances.py
Comment (kaxil) [reply tests/api_fastapi/execution_api/routes/test_task_instances.py:486]: Although, you don't have to wait for that PR to be merged, you can merge this and I can handle any changes in https://github.com/apache/airflow/pull/45002

File: task_sdk/tests/execution_time/test_supervisor.py
Comment (amoghrajesh) [reply task_sdk/tests/execution_time/test_supervisor.py]: Let me nuke it by addign a comment.

File: tests/api_fastapi/execution_api/routes/test_task_instances.py
Comment (amoghrajesh) [reply tests/api_fastapi/execution_api/routes/test_task_instances.py:486]: Alright, will leave a TODO

**Comment by amoghrajesh** [comment]
@kaxil I will merge this PR once the CI is green to unblock the other PRs which are based on this branch

**Pull request merged**
